From 04ea586ddfc5f306aadcddc0b1352cdd9eb0a296 Mon Sep 17 00:00:00 2001
From: pdiopX <piotrx.diop@intel.com>
Date: Tue, 2 Jul 2013 10:50:42 +0200
Subject: Make distinction between stopped and not-yet-started input

BZ: 115755
[CLEAN]

There was a race condition when a client calls getInput(), but is preempted
by a second client getInput() call for the same source. In that case, after
the release of the latter, the input is given back to the first client if he
is still alive (see patch I57b4b27151475d449d503d8e12d7d57159d04d77).
In this situation the startInput() call from the first client will fail as
the policy assumes that it has already started and is active.

This fix implements a new inputDescriptor parameter, mHasStarted to distinguish
an input that has been stopped (and therefore already called startInput()) from
an input that did not yet start.

Change-Id: I658122893cd57b7e9dceb87bd350d2ac005b98c7
Orig-Change-Id: Id4a57e8c13a5061338be3754075fbd5ae00ec5c3
Category: device enablement
Domain: AudioComms-Common
Origin: internal
Upstream-Candidate: no, needs rework
Signed-off-by: pdiopX <piotrx.diop@intel.com>
---
 audio/AudioPolicyManagerBase.cpp                 |    2 +-
 include/hardware_legacy/AudioPolicyManagerBase.h |    1 +
 2 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/audio/AudioPolicyManagerBase.cpp b/audio/AudioPolicyManagerBase.cpp
index e64d54d..740b557 100644
--- a/audio/AudioPolicyManagerBase.cpp
+++ b/audio/AudioPolicyManagerBase.cpp
@@ -3470,7 +3470,7 @@ status_t AudioPolicyManagerBase::AudioOutputDescriptor::dump(int fd)
 AudioPolicyManagerBase::AudioInputDescriptor::AudioInputDescriptor(const IOProfile *profile)
     : mSamplingRate(0), mFormat((audio_format_t)0), mChannelMask((audio_channel_mask_t)0),
       mDevice(AUDIO_DEVICE_NONE), mRefCount(0),
-      mInputSource(0), mProfile(profile)
+      mInputSource(0), mProfile(profile), mHasStarted(0)
 {
 }
 
diff --git a/include/hardware_legacy/AudioPolicyManagerBase.h b/include/hardware_legacy/AudioPolicyManagerBase.h
index 57836f1..aa25278 100644
--- a/include/hardware_legacy/AudioPolicyManagerBase.h
+++ b/include/hardware_legacy/AudioPolicyManagerBase.h
@@ -307,6 +307,7 @@ protected:
             uint32_t mRefCount;                         // number of AudioRecord clients using this output
             int      mInputSource;                      // input source selected by application (mediarecorder.h)
             const IOProfile *mProfile;                  // I/O profile this output derives from
+            bool     mHasStarted;                       // to indicate that the audiorecord has started
         };
 
         // stream descriptor used for volume control
-- 
1.7.4.1

